_posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide.md

```diff
@@ -0,0 +1,170 @@
+
+## Why is HPA not scaling down?
+
+Although CPU usage spiked at startup, it quickly returned to low levels. So why isn't the HPA scaling down the number of replicas? It appears that the CPU requirement is well below the target value of `450m`, as illustrated in *Figure 1*. According to the official HPA documentation:
+
+> "If multiple metrics are specified in a HorizontalPodAutoscaler, this calculation is done for each metric, and then the largest of the desired replica counts is chosen."
+
+This indicates that the issue now lies with memory usage. Grafana shows us that memory usage has remained constant after the scaling. According to *Figure 2*, the expected number of replicas should be just 1. However, since the HPA previously increased our replicas to 3, when we view the same plot with the number of replicas on the x-axis, it reveals the following:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/memory_stairs.png --alt HPA memory current replicas vs expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 5. HPA memory current replicas vs expected replicas*
+{: refdef}
+
+With the current memory usage, the HPA behaves like the function $f(x) = x$, preventing the number of replicas from scaling down. This is why we continuously see three pods running, even though the pods are not receiving much traffic.
+
+## What can we do?
+
+We have a couple of options to address this problem. For instance, we could change the memory and CPU targets in the HPA settings. Increasing the CPU requirements could help us to avoid the initial spike in replicas and prevent the staircase pattern. On the other hand if we increase the memory requirements we could avoid the bottleneck and scale down the number of replicas. However, this isn't a permanent solution because if our application's memory use fluctuates, we could encounter the same issue again. Instead, we should tackle the root cause and for this we need to differentiate between the fixed and variable usage of our metrics.
+
+The initial spike in CPU usage was due to the startup process, which is a fixed usage. As we could see before, adding more replicas doesn't alleviate the problem because the usage is related with the startup process and not the traffic. Conversely, when requests arrive at our service, the CPU usage can change, allowing the HPA to scale up or scale down the number of replicas. This represents variable usage.
+
+The memory usage always stays the same, no matter how many pods are running or how much traffic there is. This is called fixed usage. Because of this, the HPA behaves like a function where $f(x) = x$. Therefore, adjusting the number of pods based on memory usage doesn't work well because the memory doesn't vary with the traffic. The best strategy is to **remove the memory metric from the HPA** settings and scale only based on the CPU metric.
```

Comment:
It's a bit weird to kind of define the term fixed usage here because you used in in the previous paragraph already